### ISSUE TITLE
Modify ec ratings system

### ISSUE THREAD

**Issue by bnchdrff** [ESDet/esdportal_profile#8]
## ESD Early Childhood Overall Ratings

* [x] Add three columns to ec_state_ratings:
  * `state_points` (supercedes `esd_el_2015.SAS_points`) (NOT NULL)
  * `total_points` (supercedes `esd_el_2015.total_points`) (NOT NULL)
  * `overall_rating` (supercedes `esd_el_2015.overall_rating`) (NOT NULL)
* [x] Modify `sort_special=ec_total_pts` and `filter_special=esd_el_2015_highscore` API arguments to use `ec_state_ratings.total_points` for sort & `ec_state_ratings.overall_rating` for filtering
* [ ] Modify scorecard to use `ec_state_ratings.total_points` & `ec_state_ratings.overall_rating` instead of the corresponding `esd_el_2015` fields, and also change `filter_special=esd_el_2015_highscore` to `filter_special=esd_el_highscore`
* [x] Write an import script to process `ec_state_ratings` CSV & apply the Rating Logic (as Drush task maybe, or a form on admin backend?)

## Rating logic

1. When ec_state_ratings is updated, ec_state_ratings.state_points updates according to the following rules: 
    1. If ec_state_ratings.PublishedRating = Empty, ec_state_ratings.state_points = 0 
    2. If ec_state_ratings.PublishedRating = 3, ec_state_ratings.state_points = 2 
    3. If ec_state_ratings.PublishedRating = 4, ec_state_ratings.state_points = 4 
    4. If ec_state_ratings.PublishedRating = 5, ec_state_ratings.state_points = 6 
    5. If ec_state_ratings.PublishedRating = In progress, award points based on most recent ec_state_ratings.PublishedRating 

2. To receive a medal: esd_el_2015.site_visit_points &gt; 0 AND esd_el_2015.staff_points &gt; 0 AND esd_el_2015.SAS_points &gt;0 
    1. If the above criteria is not met, esd_el_2015.overall_rating = Below Bronze [END RATING] 
    2. If the above criteria is met, then continue to 3. 

3. esd_el_2015.total_points = esd_el_2015.site_visit_points + esd_el_2015.staff_points + esd_el_2015.SAS_points 
    1. If esd_el_2015.total_points &gt;= 13, esd_el_2015.overall_rating = Gold [END RATING] 
    2. If esd_el_2015.total_points &gt;= 10 AND &lt;13, esd_el_2015.overall_rating = Silver [END RATING] 
    3. If esd_el_2015.total_points &gt;= 5 AND &lt;10, esd_el_2015.overall_rating = Bronze [END RATING] 
    4. In all other cases, esd_el_2015.overall_rating = Below Bronze [END RATING]

see https://docs.google.com/document/d/1nGcplm5szkhMSbfN7FVrwdiYBxAmzFGJOIAvktVR7CI/edit#heading=h.ldykp0ahylwq for original discussion.

**Comment by bnchdrff** [comment]
this seems to be handled in the frontend. moving this to the post-launch milestone.

**Comment by bnchdrff** [comment]
See https://docs.google.com/document/d/1nGcplm5szkhMSbfN7FVrwdiYBxAmzFGJOIAvktVR7CI/edit#heading=h.ldykp0ahylwq for new specifications!

**Comment by bnchdrff** [comment]
*note* renaming filter_special `has_esd_el_2015_highscore` to `has_esd_el_highscore` because ratings are continuous & rolling now!

**Comment by bnchdrff** [comment]
Ready for scorecard implmentation, @uniosx -- see note in https://github.com/ESDet/esdportal_profile/issues/8#issuecomment-123060891 about renaming the filter, and also see issue description about getting the scores from ec_state_ratings now. http://ec-scorecard-api-3mhev6qb5ihtc.us.platform.sh/ has this new code & data.

**Comment by bnchdrff** [comment]
e.g. https://ec-scorecard-api-3mhev6qb5ihtc.us.platform.sh/api/1.0/ecs.json?sort_by_special=ec_total_pts&sort_order_special=DESC&includes=most_recent_ec_state_rating&filter_special=esd_el_highscore

**Comment by bnchdrff** [comment]
@anderoy and @uniosx status check -- let me know if/when this should end up on the live api! :ghost:

**Comment by anderoy** [comment]
I am excited that this is nearing completion, but I a bit nervous about how the ratings are calculated after importing the new data. It will be important to know which programs have changed after each import so that we can do a quick spot check to make sure everything is being displayed correctly, and so that we can inform program directors of any changes to their scores.

**Comment by bnchdrff** [comment]
launching this right now... @anderoy you said this is good to go, hope it doesn't break the scorecard!

**Comment by bnchdrff** [comment]
the old devsite is dead; long live 8-ecratings-3mhev6qb5ihtc.us.platform.sh

**Comment by anderoy** [comment]
Unfortunately it looks like this isn't working properly. I am in the process of diving in and evaluating our data from all angles and this one is going to need some work. I think the issues are small and can be resolves somewhat easily.

Before making too many assumptions about what is going wrong, it would be helpful to know where the data id being pulled from and when it is happening.

Right now, at least 14 programs have a different State "PublishedRating" than before and it doesn't seem like that score is being reflected in the ec_state_ratings. This will have a minor affect on the overall medal ratings, but it does bump one program to a Gold!!

Issue 1: Updated PublishedRating not translating into new state_points and so on.
820245070 - Roberto Clemente Learning Academy has a PublishedRating of 4 which is correctly reflected in `ec_state_ratings.PublishedRating` but that isn't properly making its way to a recalculated `state_points`, `total_points`, and ultimately `overall_rating`

Issue 2: Some new ratings not reflected in the PublishedRating field (this may be a result of when and how often the data is being pulled.
820263055 - L.A.C.C. Child Care Academy should now have a 4 but it is showing up as a 3 in `ec_state_ratings`

This is something we will need to get resolved before launch.

**Comment by bnchdrff** [comment]
It looks like there might be some data type issues with the transformation script; I'll work on this.

**Comment by bnchdrff** [comment]
Semi-bad news, though -- the worklifesystems site has changed & the downloader doesn't work anymore. We might need to settle for stale data. Have you looked into how to download from them recently, @anderoy ?

**Comment by bnchdrff** [comment]
There's also a chance that I never actually ran the migration when we deployed this to production -- it was sitting up for review for a few months & eventually we had to deploy it due to some dependencies. I'll check on this tomorrow, as well... I'm a little rusty with this part of the system & need to bring myself back up to speed.

**Comment by bnchdrff** [comment]
For reference, these are direct links to the two results Eric mentioned:

https://portal.excellentschoolsdetroit.org/api/1.0/ecs/4422.json?flatten_fields=true&includes=esd_el_2015,most_recent_ec_state_rating

https://portal.excellentschoolsdetroit.org/api/1.0/ecs/3719.json?flatten_fields=true&includes=esd_el_2015,most_recent_ec_state_rating

copy/pasted & formatted: https://gist.github.com/bnchdrff/7ae008f096ba2ae26d2a

**Comment by anderoy** [comment]
Shaking a fist at Work Life Systems. To download the data I would go https://stage.worklifesystems.com/parent/4 and login with my email and password. Click export tab, select Macomb, Oakland, and Wayne counties and click export. That will give you a list of all programs.

**Comment by bnchdrff** [comment]
Fixed the bug & added corrected data to our `50-proximity` environment. Do these scores make sense?

Roberto Clemente: https://50-proximity-3mhev6qb5ihtc.us.platform.sh/api/1.0/ecs/4422.json?flatten_fields=true&includes=esd_el_2015,most_recent_ec_state_rating

L.A.C.C.: https://50-proximity-3mhev6qb5ihtc.us.platform.sh/api/1.0/ecs/3719.json?flatten_fields=true&includes=esd_el_2015,most_recent_ec_state_rating

**Comment by anderoy** [comment]
I'll check it out in the morning. Thanks for tackling this. Is it set up in
a way that I will be able to update the data easily?

**Comment by anderoy** [comment]
I think maybe the original documentation caused something quite important to be lost in translation, perhaps it was my fault for not being clear, so now that I see the error, I think a modification of the logic will resolve this. Step 1 in the "Rating Logic" at the top of the issue is good. So here is the remaining issue:
- `total_points` and `overall_rating` are not being properly updated in `ec_state_ratings`

Let me try again with the logic by rewording/fixing steps 2 and 3 **changes in bold**:
 
2. To receive a medal: `esd_el_2015.site_visit_points` > 0 AND `esd_el_2015.staff_points` > 0 AND **`ec_state_ratings.state_points` >0**

- If the above criteria is not met, `**ec_state_ratings**.overall_rating` = Below Bronze [END RATING]
- If the above criteria is met, then continue to 3.

3. `**ec_state_ratings**.total_points = esd_el_2015.site_visit_points + esd_el_2015.staff_points + **ec_state_ratings.state_points**`

- If *`*ec_state_ratings**.total_points` >= 13, **ec_state_ratings**.overall_rating = Gold [END RATING]
- If `**ec_state_ratings**.total_points` >= 10 AND <13, `**ec_state_ratings**.overall_rating` = Silver [END RATING]
- If `**ec_state_ratings**.total_points` >= 5 AND <10, `**ec_state_ratings**.overall_rating` = Bronze [END RATING]
- In all other cases, **ec_state_ratings**.overall_rating = `**esd_hs_2015.overall_rating**` [END RATING]

Attached pdf shows the error in the L.A.C.C. example from above.
[ec_state_ratings.pdf](https://github.com/ESDet/esdportal_profile/files/138294/ec_state_ratings.pdf)

**Comment by anderoy** [comment]
With that fixed, just over a dozen programs will have their total_points modified, but only one program will actually see the change in their overall_rating. But it is a big one. After this gets fixed Iris Center (info below) will be our second Gold rated program!

"rating_id": "7033",
"esd_ec_id": "820291487",

**Comment by anderoy** [comment]
@bnchdrff any status on this one?

**Comment by bnchdrff** [comment]
I've modified the script to fit the refined requirements -- you can preview data on the devsite:

https://50-proximity-3mhev6qb5ihtc.us.platform.sh/api/1.0/ecs/3719.json?flatten_fields=true&includes=most_recent_ec_state_rating

http://50-proximity-3mhev6qb5ihtc.us.platform.sh/api/1.0/ecs/3520.json?flatten_fields=true&includes=most_recent_ec_state_rating

**Comment by anderoy** [comment]
The total_points field is still not calculating correctly.

**Comment by anderoy** [comment]
Also want to make sure that if there are any other places that the total_points or overall_rating is used (i.e. the ranking query) that these new values are used.

**Comment by bnchdrff** [comment]
Can you give examples?

**Comment by anderoy** [comment]
In the top example 3719, I believe total_points should = 7

In the bottom example 3520, total_points should = 13 (which would be required for the Gold rating, so I am unsure how that is being calculated and why it isn't showing up in total_points)

**Comment by bnchdrff** [comment]
I need to modify the importer -- it's currently keeping the old
(pre-converted) total_points, but the converted total points are indeed
being used for the badge name, hence the Gold.

**Comment by bnchdrff** [comment]
OK, fixed the total_points thing.

**Comment by anderoy** [comment]
Nice. Checked and Verified. :white_check_mark: :+1:

**Comment by anderoy** [comment]
Looks like the changes here were lost somewhere along the lines.
